validator/enforce_labels_test.rego

```diff
@@ -0,0 +1,201 @@
+# limitations under the License.
+#
+
+package templates.gcp.GCPResourceLabelsV1
+
+import data.validator.gcp.lib as lib
+
+# Importing the test data 
+import data.test.fixtures.enforce_labels.assets.buckets as fixture_buckets
+import data.test.fixtures.enforce_labels.assets.compute.disks as fixture_compute_disks
+import data.test.fixtures.enforce_labels.assets.compute.images as fixture_compute_images
+import data.test.fixtures.enforce_labels.assets.compute.instances as fixture_compute_instances
+import data.test.fixtures.enforce_labels.assets.compute.snapshots as fixture_compute_snapshots
+import data.test.fixtures.enforce_labels.assets.projects as fixture_projects
+
+import data.test.fixtures.enforce_labels.constraints as fixture_constraints
+
+# Find all violations on our test cases
+find_all_violations[violation] {
+	resources := data.resources[_]
+	constraint := data.test_constraints[_]
+	issues := deny with input.asset as resources
+		 with input.constraint as constraint
+
+	violation := issues[_]
+}
+
+project_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_projects
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+bucket_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_buckets
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+compute_instance_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_compute_instances
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+compute_image_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_compute_images
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+compute_disk_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_compute_disks
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+compute_snapshot_violations[violation] {
+	constraints := [fixture_constraints.require_labels]
+	violations := find_all_violations with data.resources as fixture_compute_snapshots
+		 with data.test_constraints as constraints
+
+	violation := violations[_]
+}
+
+##### Testing for projects
+
+# # Confirm four violations were found for all projects
+test_enforce_label_projects_violates_project_four {
+	violations := project_violations
+	count(violations) == 4
+	violation := violations[_]
+	is_string(violation.msg)
+	is_object(violation.details)
+}
+
+# confirm which 3 projects are in violation
+test_enforce_label_projects_violates_project_basic {
```

Comment:
If we're looking for 3 projects, why do we count 4 above?

Reply:
we have 3 projects with violations, one of which has 2 violations (one violation per missing label).

so the first test checks that we do have 4 violations, the second one test that the right 3 bad projects are the ones flagged. This is the same for all the other tests.

Reply:
Got it, might help to have a clarifying comment.